File: src/TGManager/Services/Bundle.cs
using System.IO;
using System.IO.Compression;
using System.Net.Http;

namespace TGManager.Services;

public readonly record struct DownloadProgress(long Read, long Total)
{
    public double Fraction => Total > 0 ? Math.Clamp((double)Read / Total, 0, 1) : 0;
}

public static class Bundle
{
    public const string TelegramUrl = "https://telegram.org/dl/desktop/win64_portable";
    public const string ProxychainsUrl =
        "https://github.com/shunf4/proxychains-windows/releases/download/0.6.8/proxychains_0.6.8_win32_x64.zip";

    static readonly HttpClient Http = new()
    {
        // Таймаут — только на заголовки; тело качаем потоком и отменяем через CancellationToken.
        Timeout = TimeSpan.FromSeconds(60),
        DefaultRequestHeaders = { { "User-Agent", "TGManager/1.2 (+https://github.com/scarrymany/tg-manager)" } },
    };

    public static bool TelegramReady() => File.Exists(Paths.TelegramExe);
    public static bool ProxychainsReady() => File.Exists(Paths.ProxychainsExe);

    public static async Task DownloadTelegram(Action<string> log, Action<DownloadProgress>? progress = null, CancellationToken ct = default)
    {
        Paths.EnsureDirs();
        Directory.CreateDirectory(Paths.TelegramDir);
        var tmp = Path.Combine(Path.GetTempPath(), "tgman-" + Guid.NewGuid().ToString("N")[..8]);
        Directory.CreateDirectory(tmp);
        try
        {
            var zip = Path.Combine(tmp, "telegram.zip");
            log("Скачиваю официальный Telegram Desktop (win64 portable)…");
            await Download(TelegramUrl, zip, log, "Telegram", progress, ct);
            log("Распаковываю…");
            var extract = Path.Combine(tmp, "extract");
            await Task.Run(() => ZipFile.ExtractToDirectory(zip, extract, overwriteFiles: true), ct);
            var exe = FindFile(extract, "Telegram.exe");
            if (exe is null) throw new InvalidOperationException("В архиве не найден Telegram.exe");
            var srcDir = Path.GetDirectoryName(exe)!;
            await Task.Run(() =>
            {
                foreach (var item in Directory.GetFileSystemEntries(srcDir))
                {
                    var dest = Path.Combine(Paths.TelegramDir, Path.GetFileName(item));
                    if (Directory.Exists(item))
                    {
                        if (Directory.Exists(dest)) Directory.Delete(dest, true);
                        CopyDir(item, dest);
                    }
                    else File.Copy(item, dest, true);
                }
                if (!File.Exists(Paths.TelegramExe))
                    File.Copy(exe, Paths.TelegramExe, true);
            }, ct);
            log("✓ Готово: " + Paths.TelegramExe);
        }
        finally
        {
            try { Directory.Delete(tmp, true); } catch { /* ignore */ }
        }
    }

    public static async Task DownloadProxychains(Action<string> log, Action<DownloadProgress>? progress = null, CancellationToken ct = default)
    {
        Paths.EnsureDirs();
        Directory.CreateDirectory(Paths.ProxychainsDir);
        var tmp = Path.Combine(Path.GetTempPath(), "tgman-pc-" + Guid.NewGuid().ToString("N")[..8]);
        Directory.CreateDirectory(tmp);
        try
        {
            var zip = Path.Combine(tmp, "pc.zip");
            log("Скачиваю прокси-обёртку для Windows (ProxyChains)…");
            await Download(ProxychainsUrl, zip, log, "ProxyChains", progress, ct);
            log("Распаковываю…");
            await Task.Run(() => ZipFile.ExtractToDirectory(zip, Paths.ProxychainsDir, overwriteFiles: true), ct);
            if (!File.Exists(Paths.ProxychainsExe))
            {
                var found = FindFile(Paths.ProxychainsDir, "proxychains_win32_x64.exe")
                            ?? FindFile(Paths.ProxychainsDir, "proxychains.exe");
                if (found is null) throw new InvalidOperationException("В архиве не найден proxychains_win32_x64.exe");
                File.Copy(found, Paths.ProxychainsExe, true);
            }
            log("✓ Готово: " + Paths.ProxychainsExe);
        }
        finally
        {
            try { Directory.Delete(tmp, true); } catch { /* ignore */ }
        }
    }

    static async Task Download(string url, string dest, Action<string> log, string label,
        Action<DownloadProgress>? progress, CancellationToken ct)
    {
        using var resp = await Http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct);
        resp.EnsureSuccessStatusCode();
        var total = resp.Content.Headers.ContentLength ?? 0;
        await using var input = await resp.Content.ReadAsStreamAsync(ct);
        await using var output = File.Create(dest);
        var buf = new byte[256 * 1024];
        long read = 0;
        var lastPct = -1;
        var lastLogKb = -1L;
        while (true)
        {
            var n = await input.ReadAsync(buf, ct);
            if (n == 0) break;
            await output.WriteAsync(buf.AsMemory(0, n), ct);
            read += n;
            progress?.Invoke(new DownloadProgress(read, total));
            if (total > 0)
            {
                var pct = (int)(read * 100 / total);
                if (pct / 10 != lastPct / 10 || pct == 100)
                {
                    lastPct = pct;
                    log($"↓ {label}: {pct}% ({read / (1024 * 1024)} / {total / (1024 * 1024)} МБ)");
                }
            }
            else
            {
                var kb = read / 1024;
                if (kb / 2048 != lastLogKb / 2048)
                {
                    lastLogKb = kb;
                    log($"↓ {label}: {kb} КБ");
                }
            }
        }
        if (total > 0 && read < total)
            throw new IOException($"Скачано {read} из {total} байт — соединение оборвалось.");
    }

    static string? FindFile(string root, string name)
    {
        foreach (var f in Directory.EnumerateFiles(root, name, SearchOption.AllDirectories))
            return f;
        return null;
    }

    static void CopyDir(string src, string dest)
    {
        Directory.CreateDirectory(dest);
        foreach (var file in Directory.GetFiles(src))
            File.Copy(file, Path.Combine(dest, Path.GetFileName(file)), true);
        foreach (var dir in Directory.GetDirectories(src))
            CopyDir(dir, Path.Combine(dest, Path.GetFileName(dir)));
    }
}

File: tgmanager/bundle.py
"""Скачивание переносного Telegram и Windows-обёртки прокси (рядом с программой)."""
from __future__ import annotations

import os
import shutil
import tempfile
import zipfile
from typing import Callable
from urllib.request import Request, urlopen

from . import paths

LogFn = Callable[[str], None]

TELEGRAM_URL = "https://telegram.org/dl/desktop/win64_portable"
PROXYCHAINS_URL = (
    "https://github.com/shunf4/proxychains-windows/releases/download/"
    "0.6.8/proxychains_0.6.8_win32_x64.zip"
)

_UA = "TGManager/1.0 (+https://github.com/scarrymany/tg-manager)"


def _log(log: LogFn | None, msg: str) -> None:
    if log:
        log(msg)


def _download(url: str, dest: str, log: LogFn | None = None, label: str = "файл") -> None:
    req = Request(url, headers={"User-Agent": _UA})
    with urlopen(req, timeout=60) as resp:
        total = int(resp.headers.get("Content-Length") or 0)
        read = 0
        chunk = 256 * 1024
        with open(dest, "wb") as out:
            while True:
                buf = resp.read(chunk)
                if not buf:
                    break
                out.write(buf)
                read += len(buf)
                if total:
                    pct = int(read * 100 / total)
                    _log(log, f"↓ {label}: {pct}% ({read // (1024 * 1024)} / {total // (1024 * 1024)} МБ)")
                else:
                    _log(log, f"↓ {label}: {read // 1024} КБ")


def _find_file(root: str, names: tuple[str, ...], max_depth: int = 4) -> str | None:
    want = {n.lower() for n in names}
    root = os.path.abspath(root)
    for dirpath, dirnames, filenames in os.walk(root):
        rel = os.path.relpath(dirpath, root)
        depth = 0 if rel == "." else rel.count(os.sep) + 1
        if depth > max_depth:
            dirnames[:] = []
            continue
        for fn in filenames:
            if fn.lower() in want:
                return os.path.join(dirpath, fn)
    return None


def telegram_ready() -> bool:
    paths.refresh()
    return os.path.isfile(paths.BUNDLED_TELEGRAM_BIN)


def download_telegram(log: LogFn | None = None) -> str:
    """Качает официальный portable Telegram x64 в ./telegram. Возвращает путь к exe."""
    paths.ensure_dirs()
    dest_dir = paths.BUNDLED_TELEGRAM_DIR
    os.makedirs(dest_dir, exist_ok=True)
    tmp = tempfile.mkdtemp(prefix="tgman-dl-")
    try:
        zip_path = os.path.join(tmp, "telegram.zip")
        _log(log, "Скачиваю официальный Telegram Desktop (win64 portable)…")
        _download(TELEGRAM_URL, zip_path, log, "Telegram")
        _log(log, "Распаковываю…")
        extract = os.path.join(tmp, "extract")
        os.makedirs(extract, exist_ok=True)
        with zipfile.ZipFile(zip_path) as zf:
            zf.extractall(extract)
        exe = _find_file(extract, ("Telegram.exe", "telegram.exe"))
        if not exe:
            raise RuntimeError("В архиве не найден Telegram.exe")
        src_dir = os.path.dirname(exe)
        # копируем всё содержимое папки с exe (Updater.exe и т.д.)
        for name in os.listdir(src_dir):
            s = os.path.join(src_dir, name)
            d = os.path.join(dest_dir, name)
            if os.path.isdir(s):
                if os.path.exists(d):
                    shutil.rmtree(d, ignore_errors=True)
                shutil.copytree(s, d)
            else:
                shutil.copy2(s, d)
        final = os.path.join(dest_dir, "Telegram.exe")
        if not os.path.isfile(final):
            # на всякий случай — прямое копирование exe
            shutil.copy2(exe, final)
        _log(log, f"✓ Готово: {final}")
        return final
    finally:
        shutil.rmtree(tmp, ignore_errors=True)


def proxychains_ready() -> bool:
    paths.refresh()
    return os.path.isfile(paths.proxychains_exe())


def download_proxychains(log: LogFn | None = None) -> str:
    """Качает proxychains-windows x64 в ./tools/proxychains. Возвращает путь к exe."""
    paths.ensure_dirs()
    dest_dir = paths.PROXYCHAINS_DIR
    os.makedirs(dest_dir, exist_ok=True)
    tmp = tempfile.mkdtemp(prefix="tgman-pc-")
    try:
        zip_path = os.path.join(tmp, "pc.zip")
        _log(log, "Скачиваю прокси-обёртку для Windows (ProxyChains)…")
        _download(PROXYCHAINS_URL, zip_path, log, "ProxyChains")
        _log(log, "Распаковываю…")
        with zipfile.ZipFile(zip_path) as zf:
            zf.extractall(dest_dir)
        exe = paths.proxychains_exe()
        if not os.path.isfile(exe):
            found = _find_file(dest_dir, ("proxychains_win32_x64.exe", "proxychains.exe"))
            if not found:
                raise RuntimeError("В архиве не найден proxychains_win32_x64.exe")
            if os.path.abspath(found) != os.path.abspath(exe):
                shutil.copy2(found, exe)
        _log(log, f"✓ Готово: {exe}")
        return exe
    finally:
        shutil.rmtree(tmp, ignore_errors=True)
